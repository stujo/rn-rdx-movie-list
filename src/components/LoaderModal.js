// @flow

import React, { Component } from 'react';
import {
    ActivityIndicator,
    Dimensions,
    Modal,
    StyleSheet,
    View,
} from 'react-native';

// https://medium.com/@kelleyannerose/react-native-activityindicator-for-a-quick-easy-loading-animation-593c06c044dc

type Props = {
    loading: boolean
};

export const LoaderModal = (props: Props) => {
    const {
        loading,
        ...attributes
    } = props;

    const { width, height } = Dimensions.get('window')
    const indicatorSize = Math.min.apply(undefined, [width, height]) / 2
    const sizing = { width: indicatorSize, height: indicatorSize }

    return (
        <Modal
            transparent={true}
            animationType={'none'}
            visible={loading}>
            <View style={styles.modalBackground}>
                <View style={[styles.activityIndicatorWrapper, sizing]} >
                    <ActivityIndicator
                        size="large"
                        color="#EE7777"
                        animating={loading} />
                </View>
            </View>
        </Modal >
    )
}

const styles = StyleSheet.create({
    modalBackground: {
        flex: 1,
        alignItems: 'center',
        flexDirection: 'column',
        justifyContent: 'space-around',
        backgroundColor: '#00000040'
    },
    activityIndicatorWrapper: {
        backgroundColor: '#FFFFFF',
        borderRadius: 10,
        display: 'flex',
        alignItems: 'center',
        justifyContent: 'space-around'
    },
});

export default LoaderModal;