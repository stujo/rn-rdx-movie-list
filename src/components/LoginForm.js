import React from 'react';
import { Component } from 'react'
import { connect } from 'react-redux';
import { Button, StyleSheet, Text, TextInput, View } from 'react-native';
import { attemptLogin, attemptLogout } from '../services/Authentication/actions'
import LoaderModal from '../components/LoaderModal';
import StyleConstants from '../StyleConstants';


const fieldBasics = { ...StyleConstants.textInput, height: 30, width: 140 }

const styles = StyleSheet.create({
    formContainer: { ...StyleConstants.formContainer },
    inputContainer: { ...StyleConstants.inputContainer },
    textField: { ...fieldBasics },
    invalidTextField: { ...fieldBasics, ...StyleConstants.invalidInput },
    buttonsContainer: { ...StyleConstants.buttonsContainer },
    submitButton: { ...StyleConstants.submitButton },
});

class LoginForm extends Component {

    hasError = (field) => {
        return !!this.state[field].error
    }

    onUsernameUpdate = (value) => {
        let error = undefined;
        if (!value || value.length < 5) {
            error = "Too Short!"
        }
        this.setState({ username: { value, error } })
        return error === undefined;
    }

    onPasswordUpdate = (value) => {
        let error = undefined;
        if (!value || value.length < 5) {
            error = "Too Short!"
        }
        this.setState({ password: { value, error } })
        return error === undefined;
    }

    validateForm = () => {
        let valid = true
        valid = this.onUsernameUpdate(this.state.username.value) && valid
        valid = this.onPasswordUpdate(this.state.password.value) && valid
        return valid
    }

    validateFormAndSend = () => {
        if (this.validateForm()) {
            this.props.onAttemptLogin({
                username: this.state.username.value,
                password: this.state.password.value
            })
        }
    }

    fieldStyle = (field) => {
        if (this.hasError(field)) {
            return styles.invalidTextField
        } else {
            return styles.textField
        }
    }

    constructor(props) {
        super(props)
        this.state = {
            username: { value: 'demo123', error: undefined },
            password: { value: 'passwd', error: undefined },
        }
    }

    render() {
        return (
            <View style={styles.formContainer}>
                <LoaderModal loading={this.props.busy} />
                <View style={styles.inputContainer}>
                    <TextInput
                        placeholder="Username..."
                        autoCapitalize='none'
                        autoCorrect={false}
                        style={this.fieldStyle("username")}
                        onChangeText={this.onUsernameUpdate} value={this.state.username.value}
                        onSubmitEditing={this.validateFormAndSend}
                    />
                </View>
                <View style={styles.inputContainer}>
                    <TextInput
                        placeholder="Password..."
                        style={this.fieldStyle("password")}
                        autoCapitalize='none'
                        autoCorrect={false}
                        secureTextEntry={true}
                        onSubmitEditing={this.validateFormAndSend}
                        onChangeText={this.onPasswordUpdate} value={this.state.password.value}
                    />
                </View>
                <View style={styles.buttonsContainer}>
                    <Button
                        style={styles.submitButton}
                        containerViewStyle={{ width: '100%', marginLeft: 0 }}
                        title="Login"
                        onPress={this.validateFormAndSend} />
                </View>
            </View>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        busy: state.authenticationService ? state.authenticationService.isFetching : false
    }
};

const mapDispatchToProps = (dispatch) => {
    return {
        onAttemptLogin: (creds) => dispatch(attemptLogin(creds)),
    }
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(LoginForm);