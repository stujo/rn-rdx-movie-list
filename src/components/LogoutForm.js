import React from 'react';
import { Component } from 'react'
import { connect } from 'react-redux';
import { Button, StyleSheet, View } from 'react-native';
import { attemptLogout } from '../services/Authentication/actions'
import LoaderModal from '../components/LoaderModal';
import StyleConstants from '../StyleConstants';

const styles = StyleSheet.create({
    formContainer: { ...StyleConstants.formContainer },
    buttonsContainer: { ...StyleConstants.buttonsContainer },
    submitButton: { ...StyleConstants.submitButton },
});

class LogoutForm extends Component {

    validateForm = () => {
        let valid = true
        return valid
    }

    validateFormAndSend = () => {
        if (this.validateForm()) {
            this.props.onAttemptLogout({})
        }
    }

    render() {
        return (
            <View style={styles.formContainer}>
                <LoaderModal loading={this.props.busy} />
                <View style={styles.buttonsContainer}>
                    <Button
                        style={styles.submitButton}
                        containerViewStyle={{ width: '100%', marginLeft: 0 }}
                        title="Logout"
                        onPress={this.validateFormAndSend} />
                </View>
            </View>
        );
    }
}

const mapStateToProps = (state) => {
    return {
        busy: state.authenticationService ? state.authenticationService.isFetching : false
    }
};

const mapDispatchToProps = (dispatch) => {
    return {
        onAttemptLogout: (details) => dispatch(attemptLogout(details)),
    }
}

export default connect(
    mapStateToProps,
    mapDispatchToProps
)(LogoutForm);