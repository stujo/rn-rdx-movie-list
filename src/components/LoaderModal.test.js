
import 'react-native';
import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { LoaderModal } from '../../src/components/LoaderModal';

describe('components/LoaderModal', () => {
    let renderer;

    beforeEach(() => {
        renderer = new ShallowRenderer();
    });

    it('renders without errors', () => {

        const result = renderer.render(
            <LoaderModal />,
        );
        expect(result).toMatchSnapshot();
    });
});