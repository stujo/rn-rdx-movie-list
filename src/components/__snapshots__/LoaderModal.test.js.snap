// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/LoaderModal renders without errors 1`] = `
<Component
  animationType="none"
  hardwareAccelerated={false}
  transparent={true}
  visible={true}
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#00000040",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "space-around",
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "alignItems": "center",
            "backgroundColor": "#FFFFFF",
            "borderRadius": 10,
            "display": "flex",
            "justifyContent": "space-around",
          },
          Object {
            "height": 375,
            "width": 375,
          },
        ]
      }
    >
      <ActivityIndicator
        animating={true}
        color="#EE7777"
        hidesWhenStopped={true}
        size="large"
      />
    </View>
  </View>
</Component>
`;
